packages/bull-vault/src/AuctionBull.sol

```diff
@@ -0,0 +1,565 @@
+                // check that orders are in order
+                if (!_isDepositingInCrab) {
+                    require(currentPrice >= prevPrice, "AB7");
+                } else {
+                    require(currentPrice <= prevPrice, "AB8");
+                }
+                prevPrice = currentPrice;
+
+                _transferFromOrder(_orders[i], remainingAmount, _clearingPrice);
+
+                if (remainingAmount > _orders[i].quantity) {
+                    remainingAmount = remainingAmount.sub(_orders[i].quantity);
+                } else {
+                    break;
+                }
+            }
+        }
+
+        if (_isDepositingInCrab) {
+            /**
+             * if auction depositing into crab:
+             * - calc amount of ETH needed to deposit into crab and get crabAmount
+             * - if target WETH to have in euler greater than current amount in euler, borrow USDC to buy more WETH and deposit in euler
+             * - if target WETH to have in euler less than current amount in euler, remove WETH from euler
+             * - deposit into crab, and pay auction traders wPowerPerp  
+             */
+            _executeCrabDeposit(_crabAmount, _wethTargetInEuler, _wethLimitPrice, ethInCrab);
+            {
+                uint256 remainingAmount = wPowerPerpAmount;
+                uint256 ordersLength = _orders.length;
+                for (uint256 i; i < ordersLength; ++i) {
+                    _transferToOrder(_orders[i], remainingAmount, _clearingPrice);
+                    if (remainingAmount > _orders[i].quantity) {
+                        remainingAmount = remainingAmount.sub(_orders[i].quantity);
+                    } else {
+                        break;
+                    }
+                }
+            }
+        } else {
+            IBullStrategy(bullStrategy).redeemCrabAndWithdrawWEth(_crabAmount, wPowerPerpAmount);
+
+            {
+                uint256 remainingAmount = wPowerPerpAmount;
+                uint256 ordersLength = _orders.length;
+                for (uint256 i; i < ordersLength; ++i) {
+                    _transferToOrder(_orders[i], remainingAmount, _clearingPrice);
+                    if (remainingAmount > _orders[i].quantity) {
+                        remainingAmount = remainingAmount.sub(_orders[i].quantity);
+                    } else {
+                        break;
+                    }
+                }
+            }
+
+            _rebalanceLeverageComponent(_wethTargetInEuler, _wethLimitPrice);
+        }
+    }
+
+    function _executeCrabDeposit(uint256 _crabAmount, uint256 _wethTargetInEuler, uint256 _wethLimitPrice, uint256 _ethInCrab) internal {
+        uint256 wethFromEuler;
+        uint256 ethNeededForCrab = _crabAmount.wdiv(IERC20(crab).totalSupply()).wmul(_ethInCrab);
+        uint256 wethInCollateral = IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy));
+        if (_wethTargetInEuler > wethInCollateral) {
```

Comment:
the logic here doesn't seemt to be right

we need to look at how much (ethNeededForCrab-ethReceivedFromAuction)+(_wethTargetInEuler-wethInCollateral) to know the net amount we need to buy or sell here.

Reply:
then we buy or sell ETH to get to our target total amount we need, and then allocate between ETH needed for crab and eth needed for leverage